spring-integration-ftp/src/main/java/org/springframework/integration/ftp/session/FtpRemoteFileTemplate.java

```diff
@@ -58,21 +60,33 @@ public T doInSession(Session<FTPFile> session) throws IOException {
 		});
 	}
 
+	/**
+	 * This particular FTP implementation is based on the {@link FTPClient#listNames(String)},
+	 * because not all FTP servers properly implement the {@code STAT} command and
+	 * it doesn't provide correct information for all cases.
+	 * Also, since this method is intended for only file existence checks,
+	 * it doesn't rely on the {@link Session#exists(String)}, because the last one uses
+	 * {@link FTPClient#changeWorkingDirectory(String)} as a fallback option, when
+	 * {@link FTPClient#listNames(String)} may not return the proper answer for directories
+	 * on some servers, too.
+	 * @param path the remote file path to check.
+	 * @return true or false if remote file exists or not.
+	 */
 	@Override
 	public boolean exists(final String path) {
-		return executeWithClient(new ClientCallback<FTPClient, Boolean>() {
+		return doExecuteWithClient(new ClientCallback<FTPClient, Boolean>() {
 
 			@Override
 			public Boolean doWithClient(FTPClient client) {
 				try {
-					return client.getStatus(path) != null;
+					String[] names = client.listNames(path);
+					return !ObjectUtils.isEmpty(names);
```

Comment:
No; this will break existing applications (not our code) that are using this method to check for the existence of a directory using an FTP server that properly supports STAT.

This needs to work for both files and directories, BUT, I don't want to take the overhead of 4 round trips to the server for cases where we __know__ we are looking for a file.

I don't see an easy way to provide backwards compatibility, and no performance hit, without adding more methods to the template.